app.js

```diff
@@ -304,88 +304,85 @@ module.exports = {
+            if (!continueWriting) {
             // The buffer was full, pause the queue to stop the writes until we
             // get a drain event
-            if (q && !q.paused) {
-              q.pause();
-              targetStream.once('drain', function() {
-                q.resume();
-              });
+              if (q && !q.paused) {
+                q.pause();
+                targetStream.once('drain', function() {
+                  q.resume();
+                });
+              }
             }
-          }
-        });
-        // For errors we expect, may or may not be fatal
-        addEventListener(listenerErrorIndicator, internalEE, 'error', function(err) {
-          debug('Error ' + JSON.stringify(err));
-          callback(err);
-        });
-        addEventListener(listenerErrorIndicator, internalEE, 'finished', function(obj) {
-          function emitFinished() {
-            debug('Backup complete - written ' + JSON.stringify(obj));
-            const summary = { total: obj.total };
-            ee.emit('finished', summary);
-            if (callback) callback(null, summary);
-          }
-          if (targetStream === process.stdout) {
+          });
+          // For errors we expect, may or may not be fatal
+          addEventListener(listenerErrorIndicator, internalEE, 'error', function(err) {
+            debug('Error ' + JSON.stringify(err));
+            callback(err);
+          });
+          addEventListener(listenerErrorIndicator, internalEE, 'finished', function(obj) {
+            function emitFinished() {
+              debug('Backup complete - written ' + JSON.stringify(obj));
+              const summary = { total: obj.total };
+              ee.emit('finished', summary);
+              if (callback) callback(null, summary);
+            }
+            if (targetStream === process.stdout) {
             // stdout cannot emit a finish event so use a final write + callback
-            targetStream.write('', 'utf8', emitFinished);
-          } else {
+              targetStream.write('', 'utf8', emitFinished);
+            } else {
             // If we're writing to a file, end the writes and register the
             // emitFinished function for a callback when the file stream's finish
             // event is emitted.
-            targetStream.end('', 'utf8', emitFinished);
+              targetStream.end('', 'utf8', emitFinished);
+            }
+          });
+        } else {
+          // If resuming write a newline as it's possible one would be missing from
+          // an interruption of the previous backup. If the backup was clean this
+          // will cause an empty line that will be gracefully handled by the restore.
+          if (opts.resume) {
+            targetStream.write('\n');
           }
-        });
-      });
+
+          return backupFull(backupDB, opts, targetStream, ee)
+            .then(callback); // TODO move after shallow backup is refactored
```

Comment:
This bit confuses me - what does it do?

Reply:
This calls the user supplied `callback` (which is a user supplied function when using the API or a function from the bin scripts for CLI usage). We have to have a callback to retain the existinng API contract.

Right now there is a split between shallow and full backup in `app.js`. In the shallow half is still doing stuff via an internal event emitter, including the user API callback.
In the full version we need to callback after the pipeline promise completes. That's what this is doing.

Once we do shallow pipeline as well then this will be able to move to the end of the promise chain because it will all be able to happen in a single place.

Reply:
But our `backupFull` doesn't return anything on success, it just emits `finished` and voids and we have `catch` to  call `callback` on failure, so this is no-op? Sorry, I should've elaborate on that this is the part that confuses me

Reply:
I think I finally got it, this is calling `callback(null, undefined)` on successful finish, so user can run theirs' side-effects? weird, but ok